api-server/llama-core/src/search.rs

```diff
@@ -0,0 +1,389 @@
+                    Err(e) => {
+                        let msg = e.to_string();
+                        #[cfg(feature = "logging")]
+                        error!(target: "search", "perform_search: {}", msg);
+                        return Err(LlamaCoreError::Search(format!(
+                            "When converting to a JSON object: {}",
+                            msg
+                        )));
+                    }
+                };
+            }
+        };
+
+        // start cleaning the output.
+
+        // produce SearchOutput instance with the raw results obtained from the endpoint.
+        let mut search_output: SearchOutput = match self.parse_into_results(&raw_results) {
+            Ok(search_output) => search_output,
+            Err(e) => {
+                let msg = e.to_string();
+                #[cfg(feature = "logging")]
+                error!(target: "search", "perform_search: {}", msg);
+                return Err(LlamaCoreError::Search(format!(
+                    "When calling parse_into_results: {}",
+                    msg
+                )));
+            }
+        };
+
+        // apply maximum search result limit.
+        search_output
+            .results
+            .truncate(self.max_search_results as usize);
+
+        // apply per result character limit.
+        //
+        // since the clipping only happens when split_at_checked() returns Some, the results will
+        // remain unchanged should split_at_checked() return None.
+        for result in search_output.results.iter_mut() {
+            if let Some(clipped_content) = result
+                .text_content
+                .split_at_checked(self.size_limit_per_result as usize)
+            {
+                result.text_content = clipped_content.0.to_string();
+            }
+        }
+
+        // Search Output cleaned and finalized.
+        Ok(search_output)
+    }
+    /// Perform a search and summarize the corresponding search results
+    pub async fn summarize_search<T: Serialize>(
+        &self,
+        search_input: &T,
+    ) -> Result<String, LlamaCoreError> {
+        let search_output = self.perform_search(&search_input).await?;
+
+        let summarization_prompts = self.summarization_prompts.clone().unwrap_or((
+            "The following are search results I found on the internet:\n\n".to_string(),
+            "\n\nTo sum up them up: ".to_string(),
+        ));
+
+        // the fallback context size limit for the search summary to be generated.
+        let summarize_ctx_size = self
+            .summarize_ctx_size
+            .unwrap_or((self.size_limit_per_result * self.max_search_results as u16) as usize);
+
+        summarize(search_output, summarize_ctx_size, summarization_prompts)
+    }
+}
+
+/// Summarize the search output provided
+fn summarize(
+    search_output: SearchOutput,
+    summarize_ctx_size: usize,
+    (initial_prompt, final_prompt): (String, String),
```

Comment:
Is it possible to define a struct type to hold both `initial_prompt` and `final_prompt`? It would make the function signature more concise.

Reply:
We could do that, but what purpose would it serve? This function is not public and hence not meant for end use. Unless there is another reason for making the function signature more concise?

The `summarize_search()` function is the wrapper meant for end use. Let me know what you think

Reply:
Here is the `summarize_search()` function signature for reference:

```rust
pub async fn summarize_search<T: Serialize>(
    &self,
    search_input: &T,
) -> Result<String, LlamaCoreError> {
```

Reply:
Done. Thanks.